src/base/bittorrent/abstractfilestorage.cpp

```diff
@@ -0,0 +1,140 @@
+            first.chop(QB_EXT.size());
+        if (second.endsWith(QB_EXT, Qt::CaseInsensitive))
+            second.chop(QB_EXT.size());
+        return QString::compare(first, second, CASE_SENSITIVITY) == 0;
+    }
+}
+
+void BitTorrent::AbstractFileStorage::renameFile(const QString &oldPath, const QString &newPath)
+{
+    if (!Utils::Fs::isValidFileSystemName(oldPath, true))
+        throw RuntimeError {tr("The path is invalid: '%1'.").arg(oldPath)};
+    if (!Utils::Fs::isValidFileSystemName(newPath, true))
+        throw RuntimeError {tr("The path is invalid: '%1'.").arg(newPath)};
+
+    const QString oldFilePath = Utils::Fs::toUniformPath(oldPath);
+    if (oldFilePath.endsWith(QLatin1Char {'/'}))
+        throw RuntimeError {tr("Invalid file path: '%1'.").arg(oldFilePath)};
+
+    const QString newFilePath = Utils::Fs::toUniformPath(newPath);
+    if (newFilePath.endsWith(QLatin1Char {'/'}))
+        throw RuntimeError {tr("Invalid file path: '%1'.").arg(newFilePath)};
+    if (QDir().isAbsolutePath(newFilePath))
+        throw RuntimeError {tr("Absolute path isn't allowed: '%1'.").arg(newFilePath)};
+
+    int renamingFile = -1;
+    for (int i = 0; i < filesCount(); ++i)
+    {
+        const QString path = filePath(i);
+
+        if ((renamingFile < 0) && areSameFileNames(path, oldFilePath))
+            renamingFile = i;
+
+        if (areSameFileNames(path, newFilePath))
+            throw RuntimeError {tr("The file already exists: '%1'.").arg(newFilePath)};
+    }
+
+    if (renamingFile < 0)
+        throw RuntimeError {tr("No such file: '%1'.").arg(oldFilePath)};
+
+    const auto extAdjusted = [](const QString &path, bool needExt) -> QString
+    {
+        if (path.endsWith(QB_EXT, Qt::CaseInsensitive) == needExt)
+            return path;
+
+        return (needExt ? (path + QB_EXT) : (path.left(path.size() - QB_EXT.size())));
+    };
+
+    renameFile(renamingFile, extAdjusted(newFilePath, filePath(renamingFile).endsWith(QB_EXT, Qt::CaseInsensitive)));
+}
+
+void BitTorrent::AbstractFileStorage::renameFolder(const QString &oldPath, const QString &newPath)
+{
+    if (!Utils::Fs::isValidFileSystemName(oldPath, true))
+        throw RuntimeError {tr("The path is invalid: '%1'.").arg(oldPath)};
+    if (!Utils::Fs::isValidFileSystemName(newPath, true))
+        throw RuntimeError {tr("The path is invalid: '%1'.").arg(newPath)};
+
+    const auto cleanFolderPath = [](const QString &path) -> QString
+    {
+        const QString uniformPath = Utils::Fs::toUniformPath(path);
+        return (uniformPath.endsWith(QLatin1Char {'/'}) ? uniformPath : uniformPath + QLatin1Char {'/'});
+    };
+
+    const QString oldFolderPath = cleanFolderPath(oldPath);
+    const QString newFolderPath = cleanFolderPath(newPath);
+    if (QDir().isAbsolutePath(newFolderPath))
+        throw RuntimeError {tr("Absolute path isn't allowed: '%1'.").arg(newFolderPath)};
+
+    QVector<int> renamingFiles;
```

Comment:
Looks like this is storing indexes, if so, I would rename it to `renamingFileIndexes` or `renamingIndexs`

Reply:
> `renamingIndexes`

We don't rename indexes.
I'll use `renamingFileIndexes` if it makes sense for you.